src/Components/DataHistoryExplorer/DataHistoryExplorer.tsx

```diff
@@ -39,24 +80,371 @@ const DataHistoryExplorer: React.FC<IDataHistoryExplorerProps> = (props) => {
+                            timeSeriesTwin.label !==
+                            getDefaultSeriesLabel(
+                                timeSeriesTwin.twinId,
+                                timeSeriesTwin.twinPropertyName
+                            )
+                    }
+                ]);
+            }
+        },
+        [
+            isDefined,
+            dispatch,
+            sendDataHistoryExplorerUserTelemetry,
+            getDefaultSeriesLabel,
+            getHighChartColor,
+            state.timeSeriesTwins,
+            state.selectedTimeSeriesId
+        ]
+    );
+
+    const handleOnAddSeriesClick = useCallback(
+        (calloutTargetId: string) => {
+            calloutTargetIdRef.current = calloutTargetId;
+            dispatch({
+                type:
+                    DataHistoryExplorerActionType.SET_IS_TIME_SERIES_TWIN_CALLOUT_VISIBLE,
+                payload: { isVisible: true }
+            });
+        },
+        [dispatch]
+    );
+
+    const handleOnEditSeriesClick = useCallback(
+        (seriesId: string, calloutTargetId: string) => {
+            calloutTargetIdRef.current = calloutTargetId;
+            dispatch({
+                type:
+                    DataHistoryExplorerActionType.SET_IS_TIME_SERIES_TWIN_CALLOUT_VISIBLE,
+                payload: { isVisible: true, selectedSeriesId: seriesId }
+            });
+        },
+        [dispatch]
+    );
+
+    const handleOnRemoveSeriesClick = useCallback(
+        (seriesId: string) => {
+            const selectedIdx = state.timeSeriesTwins.findIndex(
+                (tsTwin) => tsTwin.seriesId === seriesId
+            );
+            dispatch({
+                type: DataHistoryExplorerActionType.REMOVE_TIME_SERIES_TWINS,
+                payload: { seriesIdx: selectedIdx }
+            });
+            const telemetry =
+                TelemetryEvents.Tools.DataHistoryExplorer.UserAction
+                    .RemoveSeries;
+            sendDataHistoryExplorerUserTelemetry(telemetry.eventName, [
+                {
+                    [telemetry.properties.itemIndex]: selectedIdx
+                }
+            ]);
+            /**
+             * when a series is removed, also remove if from usedColors reference
+             * so that whenever adding a new series we can use these available colors first
+             * before picking the next color in Highcharts palette
+             */
+            usedSeriesColorsRef.current.splice(
+                usedSeriesColorsRef.current.findIndex(
+                    (c) =>
+                        c ===
+                        state.timeSeriesTwins.find(
+                            (tsTwin) => tsTwin.seriesId === seriesId
+                        ).chartProps.color
+                ),
+                1
+            );
+        },
+        [state.timeSeriesTwins, dispatch, sendDataHistoryExplorerUserTelemetry]
+    );
+
+    const handleOnViewModeChange = useCallback(
```

Comment:
love how you have coupled the handlers to the telemetry. This is great. Very nicely done. Love to see the instrumentation.